<a href="https://cognitiveclass.ai"><img src = "https://ibm.box.com/shared/static/9gegpsmnsoo25ikkbl4qzlvlyjbgxs5x.png" width = 400> </a>

<h1 align=center><font size = 5>Segmenting and Clustering Neighborhoods in Toronto City</font></h1>
<h2 align=center><font size = 5>Part 3: Explore and cluster the neighborhoods in Toronto </font></h2>

## Introduction

In this lab, you will learn how to convert addresses into their equivalent latitude and longitude values. Also, you will use the Foursquare API to explore neighborhoods in Toronto City. You will use the **explore** function to get the most common venue categories in each neighborhood, and then use this feature to group the neighborhoods into clusters. You will use the *k*-means clustering algorithm to complete this task. Finally, you will use the Folium library to visualize the neighborhoods in Toronto City and their emerging clusters.

Before we get the data and start exploring it, let's download all the dependencies that we will need.

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')


Libraries imported.


### Read CSV genersted in part 2

In [3]:
neighborhoods = pd.read_csv('Merge_Toronto_Neighborhoods_with_coordinates.csv')
print('Data shape=' ,neighborhoods.shape)
neighborhoods.head()

Data shape= (103, 5)


Postcode           Borough                     Neighborhood   Latitude  \
0      M3A        North York                        Parkwoods  43.753259   
1      M4A        North York                 Victoria Village  43.725882   
2      M5A  Downtown Toronto                     Harbourfront  43.654260   
3      M6A        North York  Lawrence Heights,Lawrence Manor  43.718518   
4      M7A  Downtown Toronto                     Queen's Park  43.662301   

   Longitude  
0 -79.329656  
1 -79.315572  
2 -79.360636  
3 -79.464763  
4 -79.389494

#### Use geopy library to get the latitude and longitude values of Toronto City.

In order to define an instance of the geocoder, we need to define a user_agent. We will name our agent <em>ny_explorer</em>, as shown below.

In [60]:
address = 'Toronto, CA'
#geolocator = Nominatim(user_agent="ny_explorer")
#location = geolocator.geocode(address)
latitude = 43.653963
longitude = -79.387207
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


#### Create a map of Toronto with neighborhoods superimposed on top.

In [61]:
# create map of Toronto using latitude and longitude values
map_newyork = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_newyork)  
    
map_newyork

**Folium** is a great visualization library. Feel free to zoom into the above map, and click on each circle mark to reveal the name of the neighborhood and its respective borough.

Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.

#### Define Foursquare Credentials and Version

In [62]:
CLIENT_ID = 'TVSPG3MPILEHFZ5T0WUADCTYK0PU2UAHPB2XSRDQE2EFOGG4' # your Foursquare ID
CLIENT_SECRET = 'RXIMYAIACRTJX0QTI5MIDJYOF0BMHUFTQVFGPVLEHMSVA24A' # your Foursquare Secret
VERSION = '20180605'


print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: TVSPG3MPILEHFZ5T0WUADCTYK0PU2UAHPB2XSRDQE2EFOGG4
CLIENT_SECRET:RXIMYAIACRTJX0QTI5MIDJYOF0BMHUFTQVFGPVLEHMSVA24A


#### Let's explore the first neighborhood in our dataframe.

Get the neighborhood's name.

In [63]:

neighborhoods.loc[0, 'Neighborhood']

'Parkwoods'

Get the neighborhood's latitude and longitude values.

In [64]:
neighborhood_latitude = neighborhoods.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = neighborhoods.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = neighborhoods.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of "{}" are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of "Parkwoods" are 43.7532586, -79.3296565.


#### Now, let's get the top 100 venues that are in Parkwoods within a radius of 500 meters.

First, let's create the GET request URL. Name your URL **url**.

In [72]:
# type your answer here
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius

url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=TVSPG3MPILEHFZ5T0WUADCTYK0PU2UAHPB2XSRDQE2EFOGG4&client_secret=RXIMYAIACRTJX0QTI5MIDJYOF0BMHUFTQVFGPVLEHMSVA24A&v=20180605&ll=43.7532586,-79.3296565&radius=500&limit=100'

Send the GET request and examine the resutls

In [73]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5e7513b0df2774001b77859f'},
 'response': {'warning': {'text': "There aren't a lot of results near you. Try something more general, reset your filters, or expand the search area."},
  'headerLocation': 'Parkwoods - Donalda',
  'headerFullLocation': 'Parkwoods - Donalda, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 2,
  'suggestedBounds': {'ne': {'lat': 43.757758604500005,
    'lng': -79.32343823984928},
   'sw': {'lat': 43.7487585955, 'lng': -79.33587476015072}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4e8d9dcdd5fbbbb6b3003c7b',
       'name': 'Brookbanks Park',
       'location': {'address': 'Toronto',
        'lat': 43.751976046055574,
        'lng': -79.33214044722958,
        'labeledLatLngs': 

From the Foursquare lab in the previous module, we know that all the information is in the *items* key. Before we proceed, let's borrow the **get_category_type** function from the Foursquare lab.

In [68]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Now we are ready to clean the json and structure it into a *pandas* dataframe.

In [69]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON
# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()


name         categories        lat        lng
0  Brookbanks Park               Park  43.751976 -79.332140
1    Variety Store  Food & Drink Shop  43.751974 -79.333114

And how many venues were returned by Foursquare?

In [14]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

2 venues were returned by Foursquare.


<a id='item2'></a>

## 2. Explore Neighborhoods in Toronto

#### Let's create a function to repeat the same process to all the neighborhoods in Toronto

In [45]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Now Run the above function on each neighborhood and create a new dataframe called *toronto_venues*.

In [17]:
toronto_venues = getNearbyVenues(names=neighborhoods['Neighborhood'],
                                   latitudes=neighborhoods['Latitude'],
                                   longitudes=neighborhoods['Longitude']
                                  )

toronto_venues.head()

Parkwoods
Victoria Village
Harbourfront
Lawrence Heights,Lawrence Manor
Queen's Park
Islington Avenue
Rouge,Malvern
Don Mills North
Woodbine Gardens,Parkview Hill
Ryerson,Garden District
Glencairn
Cloverdale,Islington,Martin Grove,Princess Gardens,West Deane Park
Highland Creek,Rouge Hill,Port Union
Flemingdon Park,Don Mills South
Woodbine Heights
St. James Town
Humewood-Cedarvale
Bloordale Gardens,Eringate,Markland Wood,Old Burnhamthorpe
Guildwood,Morningside,West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor,Downsview North,Wilson Heights
Thorncliffe Park
Adelaide,King,Richmond
Dovercourt Village,Dufferin
Scarborough Village
Fairview,Henry Farm,Oriole
Northwood Park,York University
East Toronto
Harbourfront East,Toronto Islands,Union Station
Little Portugal,Trinity
East Birchmount Park,Ionview,Kennedy Park
Bayview Village
CFB Toronto,Downsview East
The Danforth West,Riverdale
Design Exchange,Toro

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.753259              -79.329656   
1         Parkwoods              43.753259              -79.329656   
2  Victoria Village              43.725882              -79.315572   
3  Victoria Village              43.725882              -79.315572   
4  Victoria Village              43.725882              -79.315572   

                    Venue  Venue Latitude  Venue Longitude  \
0         Brookbanks Park       43.751976       -79.332140   
1           Variety Store       43.751974       -79.333114   
2  Victoria Village Arena       43.723481       -79.315635   
3             Tim Hortons       43.725517       -79.313103   
4               Portugril       43.725819       -79.312785   

          Venue Category  
0                   Park  
1      Food & Drink Shop  
2           Hockey Arena  
3            Coffee Shop  
4  Portuguese Restaurant

#### Let's check the size of the resulting dataframe

In [18]:
print(toronto_venues.shape)

(2248, 7)


Let's check how many venues were returned for each neighborhood

In [46]:
toronto_venues.to_csv('toronto_venues.csv',index=False)

In [20]:
toronto_venues.groupby('Neighborhood')['Venue'].count().to_frame().head()

Venue
Neighborhood                                             
Adelaide,King,Richmond                                100
Agincourt                                               5
Agincourt North,L'Amoreaux East,Milliken,Steele...      2
Albion Gardens,Beaumond Heights,Humbergate,Jame...      9
Alderwood,Long Branch                                   9

#### Let's find out how many unique categories can be curated from all the returned venues

In [21]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 274 uniques categories.


<a id='item3'></a>

## 3. Analyze Each Neighborhood

In [22]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")
toronto_onehot.head()


Accessories Store  Afghan Restaurant  Airport  Airport Food Court  \
0                  0                  0        0                   0   
1                  0                  0        0                   0   
2                  0                  0        0                   0   
3                  0                  0        0                   0   
4                  0                  0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  Antique Shop  Aquarium  Art Gallery  Art Museum  \
0                    0             0         0            0           0   
1                    0             0         0            0           0   
2                    0             0         0            0           0   
3                    0             0         0            0           0   
4                    0             0         0            0           0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Garage  \
0                    0                 0                   0            0   
1                    0                 0                   0            0   
2                    0                 0                   0            0   
3                    0                 0                   0            0   
4                    0                 0                   0            0   

   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0              0          0           0           0       0     0    0   
1              0          0           0           0       0     0    0   
2              0          0           0           0       0     0    0   
3              0          0           0           0       0     0    0   
4              0          0           0           0       0     0    0   

   Baseball Field  Baseball Stadium  Basketball Stadium  Beach  \
0               0                 0                   0      0   
1               0                 0                   0      0   
2               0                 0                   0      0   
3               0                 0                   0      0   
4               0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bike Shop  \
0                0         0           0                   0          0   
1                0         0           0                   0          0   
2                0         0           0                   0          0   
3                0         0           0                   0          0   
4                0         0           0                   0          0   

   Bistro  Board Shop  Boat or Ferry  Bookstore  Boutique  \
0       0           0              0          0         0   
1       0           0              0          0         0   
2       0           0              0          0         0   
3       0           0              0          0         0   
4       0           0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bridal Shop  \
0                     0               0        0            0   
1                     0               0        0            0   
2                     0               0        0            0   
3                     0               0        0            0   
4                     0               0        0            0   

   Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0                0         0             0              0         0   
1                0         0             0

In [23]:
# add neighborhood column back to dataframe      
toronto_onehot['Neighborhood'] =  toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns=list( toronto_onehot.columns)
fixed_columns.remove('Neighborhood')
fixed_columns =  ['Neighborhood'] + fixed_columns

toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Neighborhood  Accessories Store  Afghan Restaurant  Airport  \
0         Parkwoods                  0                  0        0   
1         Parkwoods                  0                  0        0   
2  Victoria Village                  0                  0        0   
3  Victoria Village                  0                  0        0   
4  Victoria Village                  0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0                 0                    0             0         0            0   
1                 0                    0             0         0            0   
2                 0                    0             0         0            0   
3                 0                    0             0         0            0   
4                 0                    0             0         0            0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0           0                    0                 0                   0   
1           0                    0                 0                   0   
2           0                    0                 0                   0   
3           0                    0                 0                   0   
4           0                    0                 0                   0   

   Auto Garage  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  \
0            0              0          0           0           0       0   
1            0              0          0           0           0       0   
2            0              0          0           0           0       0   
3            0              0          0           0           0       0   
4            0              0          0           0           0       0   

   Bank  Bar  Baseball Field  Baseball Stadium  Basketball Stadium  Beach  \
0     0    0               0                 0                   0      0   
1     0    0               0                 0                   0      0   
2     0    0               0                 0                   0      0   
3     0    0               0                 0                   0      0   
4     0    0               0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bike Shop  \
0                0         0           0                   0          0   
1                0         0           0                   0          0   
2                0         0           0                   0          0   
3                0         0           0                   0          0   
4                0         0           0                   0          0   

   Bistro  Board Shop  Boat or Ferry  Bookstore  Boutique  \
0       0           0              0          0         0   
1       0           0              0          0         0   
2       0           0              0          0         0   
3       0           0              0          0         0   
4       0           0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bridal Shop  \
0                     0               0        0            0   
1                     0               0        0            0   
2                     0               0        0            0   
3                     0               0        0            0   
4                     0               0        0            0   

   Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0        

And let's examine the new dataframe size.

In [24]:
toronto_onehot.shape

(2248, 274)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [25]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped.head()

Neighborhood  Accessories Store  \
0                             Adelaide,King,Richmond                0.0   
1                                          Agincourt                0.0   
2  Agincourt North,L'Amoreaux East,Milliken,Steel...                0.0   
3  Albion Gardens,Beaumond Heights,Humbergate,Jam...                0.0   
4                              Alderwood,Long Branch                0.0   

   Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                0.0      0.0                 0.0           0.0   
1                0.0      0.0                 0.0           0.0   
2                0.0      0.0                 0.0           0.0   
3                0.0      0.0                 0.0           0.0   
4                0.0      0.0                 0.0           0.0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0             0.0              0.0               0.0                 0.02   
1             0.0              0.0               0.0                 0.00   
2             0.0              0.0               0.0                 0.00   
3             0.0              0.0               0.0                 0.00   
4             0.0              0.0               0.0                 0.00   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0           0.0       0.0         0.01        0.01                  0.0   
1           0.0       0.0         0.00        0.00                  0.0   
2           0.0       0.0         0.00        0.00                  0.0   
3           0.0       0.0         0.00        0.00                  0.0   
4           0.0       0.0         0.00        0.00                  0.0   

   Asian Restaurant  Athletics & Sports  Auto Garage  Auto Workshop  \
0              0.02                 0.0          0.0            0.0   
1              0.00                 0.0          0.0            0.0   
2              0.00                 0.0          0.0            0.0   
3              0.00                 0.0          0.0            0.0   
4              0.00                 0.0          0.0            0.0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank   Bar  Baseball Field  \
0        0.0         0.0         0.0    0.02   0.0  0.03             0.0   
1        0.0         0.0         0.0    0.00   0.0  0.00             0.0   
2        0.0         0.0         0.0    0.00   0.0  0.00             0.0   
3        0.0         0.0         0.0    0.00   0.0  0.00             0.0   
4        0.0         0.0         0.0    0.00   0.0  0.00             0.0   

   Baseball Stadium  Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  \
0               0.0                 0.0    0.0              0.0       0.0   
1               0.0                 0.0    0.0              0.0       0.0   
2               0.0                 0.0    0.0              0.0       0.0   
3               0.0                 0.0    0.0              0.0       0.0   
4               0.0                 0.0    0.0              0.0       0.0   

   Beer Store  Belgian Restaurant  Bike Shop  Bistro  Board Shop  \
0    0.000000                 0.0        0.0     0.0         0.0   
1    0.000000                 0.0        0.0     0.0         0.0   
2    0.000000                 0.0        0.0     0.0         0.0   
3    0.111111                 0.0        0.0     0.0         0.0   
4    0.000000                 0.0        0.0     0.0         0.0   

   Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  \
0            0.0       0.02       0.0                  0.01            0.02   
1            0.0       0.00       0.0                  0.00            0.20   
2            0.0       0.00       0.0                  0.00            0.00   
3            0.0       0.00       0.0                  0.00            0.00   
4            0.0       0.00       0.0                  0.00            0.00   

   Brewery  Bridal Shop  Bubble Tea Shop 

#### Let's confirm the new size

In [26]:
toronto_grouped.shape

(99, 274)

### Notes: as shown Neighborhoods are 100 instead of 103 as there are three  Neighborhoods has no Venues

In [27]:
#Three Neighborhood has no Venues
a = toronto_venues['Neighborhood'].unique()
b = neighborhoods['Neighborhood'].unique()
neighborhood_has_no_venues= set(b) - set(a)
neighborhood_has_no_venues

{'Cloverdale,Islington,Martin Grove,Princess Gardens,West Deane Park',
 'Islington Avenue',
 'Newtonbrook,Willowdale',
 'Upper Rouge'}

#### Let's print each neighborhood along with the top 5 most common venues

In [28]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide,King,Richmond----
             venue  freq
0      Coffee Shop  0.07
1       Restaurant  0.05
2             Café  0.04
3  Thai Restaurant  0.04
4              Bar  0.03


----Agincourt----
                       venue  freq
0  Latin American Restaurant   0.2
1             Breakfast Spot   0.2
2                     Lounge   0.2
3               Skating Rink   0.2
4             Clothing Store   0.2


----Agincourt North,L'Amoreaux East,Milliken,Steeles East----
                       venue  freq
0                 Playground   0.5
1                       Park   0.5
2  Middle Eastern Restaurant   0.0
3                      Motel   0.0
4        Monument / Landmark   0.0


----Albion Gardens,Beaumond Heights,Humbergate,Jamestown,Mount Olive,Silverstone,South Steeles,Thistletown----
                  venue  freq
0         Grocery Store  0.22
1  Fast Food Restaurant  0.11
2          Liquor Store  0.11
3   Fried Chicken Joint  0.11
4        Sandwich Place  0.11


----Alderwood,Long B

#### Let's put that into a *pandas* dataframe

First, let's write a function to sort the venues in descending order.

In [29]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [30]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                             Adelaide,King,Richmond           Coffee Shop   
1                                          Agincourt                Lounge   
2  Agincourt North,L'Amoreaux East,Milliken,Steel...            Playground   
3  Albion Gardens,Beaumond Heights,Humbergate,Jam...         Grocery Store   
4                              Alderwood,Long Branch           Pizza Place   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                 Restaurant                  Café       Thai Restaurant   
1  Latin American Restaurant          Skating Rink        Clothing Store   
2                       Park            Donut Shop    Dim Sum Restaurant   
3        Fried Chicken Joint        Sandwich Place          Liquor Store   
4                Coffee Shop        Sandwich Place                   Pub   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                   Bar            Steakhouse                Lounge   
1        Breakfast Spot             Drugstore        Discount Store   
2                 Diner        Discount Store   Distribution Center   
3              Pharmacy  Fast Food Restaurant            Beer Store   
4          Skating Rink                  Pool                   Gym   

  8th Most Common Venue  9th Most Common Venue 10th Most Common Venue  
0                   Gym              Bookstore                 Bakery  
1   Distribution Center                Dog Run       Doner Restaurant  
2               Dog Run       Doner Restaurant              Drugstore  
3           Pizza Place  General Entertainment           Dance Studio  
4              Pharmacy    Dumpling Restaurant              Drugstore

In [31]:
toronto_grouped.head()

Neighborhood  Accessories Store  \
0                             Adelaide,King,Richmond                0.0   
1                                          Agincourt                0.0   
2  Agincourt North,L'Amoreaux East,Milliken,Steel...                0.0   
3  Albion Gardens,Beaumond Heights,Humbergate,Jam...                0.0   
4                              Alderwood,Long Branch                0.0   

   Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                0.0      0.0                 0.0           0.0   
1                0.0      0.0                 0.0           0.0   
2                0.0      0.0                 0.0           0.0   
3                0.0      0.0                 0.0           0.0   
4                0.0      0.0                 0.0           0.0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0             0.0              0.0               0.0                 0.02   
1             0.0              0.0               0.0                 0.00   
2             0.0              0.0               0.0                 0.00   
3             0.0              0.0               0.0                 0.00   
4             0.0              0.0               0.0                 0.00   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0           0.0       0.0         0.01        0.01                  0.0   
1           0.0       0.0         0.00        0.00                  0.0   
2           0.0       0.0         0.00        0.00                  0.0   
3           0.0       0.0         0.00        0.00                  0.0   
4           0.0       0.0         0.00        0.00                  0.0   

   Asian Restaurant  Athletics & Sports  Auto Garage  Auto Workshop  \
0              0.02                 0.0          0.0            0.0   
1              0.00                 0.0          0.0            0.0   
2              0.00                 0.0          0.0            0.0   
3              0.00                 0.0          0.0            0.0   
4              0.00                 0.0          0.0            0.0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank   Bar  Baseball Field  \
0        0.0         0.0         0.0    0.02   0.0  0.03             0.0   
1        0.0         0.0         0.0    0.00   0.0  0.00             0.0   
2        0.0         0.0         0.0    0.00   0.0  0.00             0.0   
3        0.0         0.0         0.0    0.00   0.0  0.00             0.0   
4        0.0         0.0         0.0    0.00   0.0  0.00             0.0   

   Baseball Stadium  Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  \
0               0.0                 0.0    0.0              0.0       0.0   
1               0.0                 0.0    0.0              0.0       0.0   
2               0.0                 0.0    0.0              0.0       0.0   
3               0.0                 0.0    0.0              0.0       0.0   
4               0.0                 0.0    0.0              0.0       0.0   

   Beer Store  Belgian Restaurant  Bike Shop  Bistro  Board Shop  \
0    0.000000                 0.0        0.0     0.0         0.0   
1    0.000000                 0.0        0.0     0.0         0.0   
2    0.000000                 0.0        0.0     0.0         0.0   
3    0.111111                 0.0        0.0     0.0         0.0   
4    0.000000                 0.0        0.0     0.0         0.0   

   Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  \
0            0.0       0.02       0.0                  0.01            0.02   
1            0.0       0.00       0.0                  0.00            0.20   
2            0.0       0.00       0.0                  0.00            0.00   
3            0.0       0.00       0.0                  0.00            0.00   
4            0.0       0.00       0.0                  0.00            0.00   

   Brewery  Bridal Shop  Bubble Tea Shop 

<a id='item4'></a>

## 4. Cluster Neighborhoods

Run *k*-means to cluster the neighborhood into 5 clusters.

In [32]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 1, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [33]:
neighborhoods.head()

Postcode           Borough                     Neighborhood   Latitude  \
0      M3A        North York                        Parkwoods  43.753259   
1      M4A        North York                 Victoria Village  43.725882   
2      M5A  Downtown Toronto                     Harbourfront  43.654260   
3      M6A        North York  Lawrence Heights,Lawrence Manor  43.718518   
4      M7A  Downtown Toronto                     Queen's Park  43.662301   

   Longitude  
0 -79.329656  
1 -79.315572  
2 -79.360636  
3 -79.464763  
4 -79.389494

In [34]:
neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                             Adelaide,King,Richmond           Coffee Shop   
1                                          Agincourt                Lounge   
2  Agincourt North,L'Amoreaux East,Milliken,Steel...            Playground   
3  Albion Gardens,Beaumond Heights,Humbergate,Jam...         Grocery Store   
4                              Alderwood,Long Branch           Pizza Place   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                 Restaurant                  Café       Thai Restaurant   
1  Latin American Restaurant          Skating Rink        Clothing Store   
2                       Park            Donut Shop    Dim Sum Restaurant   
3        Fried Chicken Joint        Sandwich Place          Liquor Store   
4                Coffee Shop        Sandwich Place                   Pub   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                   Bar            Steakhouse                Lounge   
1        Breakfast Spot             Drugstore        Discount Store   
2                 Diner        Discount Store   Distribution Center   
3              Pharmacy  Fast Food Restaurant            Beer Store   
4          Skating Rink                  Pool                   Gym   

  8th Most Common Venue  9th Most Common Venue 10th Most Common Venue  
0                   Gym              Bookstore                 Bakery  
1   Distribution Center                Dog Run       Doner Restaurant  
2               Dog Run       Doner Restaurant              Drugstore  
3           Pizza Place  General Entertainment           Dance Studio  
4              Pharmacy    Dumpling Restaurant              Drugstore

In [35]:
# add clustering labels
#neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)
neighborhoods_venues_sorted['Cluster Labels']=kmeans.labels_
toronto_merged = neighborhoods

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head(10) # check the last columns!

Postcode           Borough                     Neighborhood   Latitude  \
0      M3A        North York                        Parkwoods  43.753259   
1      M4A        North York                 Victoria Village  43.725882   
2      M5A  Downtown Toronto                     Harbourfront  43.654260   
3      M6A        North York  Lawrence Heights,Lawrence Manor  43.718518   
4      M7A  Downtown Toronto                     Queen's Park  43.662301   
5      M9A         Etobicoke                 Islington Avenue  43.667856   
6      M1B       Scarborough                    Rouge,Malvern  43.806686   
7      M3B        North York                  Don Mills North  43.745906   
8      M4B         East York   Woodbine Gardens,Parkview Hill  43.706397   
9      M5B  Downtown Toronto          Ryerson,Garden District  43.657162   

   Longitude 1st Most Common Venue   2nd Most Common Venue  \
0 -79.329656                  Park       Food & Drink Shop   
1 -79.315572          Hockey Arena            Intersection   
2 -79.360636           Coffee Shop                    Park   
3 -79.464763        Clothing Store  Furniture / Home Store   
4 -79.389494           Coffee Shop                    Park   
5 -79.532242                   NaN                     NaN   
6 -79.194353  Fast Food Restaurant             Yoga Studio   
7 -79.352188  Gym / Fitness Center     Japanese Restaurant   
8 -79.309937           Pizza Place    Gym / Fitness Center   
9 -79.378937           Coffee Shop          Clothing Store   

  3rd Most Common Venue      4th Most Common Venue 5th Most Common Venue  \
0           Yoga Studio                 Donut Shop                 Diner   
1           Coffee Shop      Portuguese Restaurant           Yoga Studio   
2                   Pub                     Bakery    Mexican Restaurant   
3     Accessories Store                Coffee Shop           Event Space   
4           Yoga Studio                 Restaurant    Italian Restaurant   
5                   NaN                        NaN                   NaN   
6            Donut Shop         Dim Sum Restaurant                 Diner   
7                  Café       Caribbean Restaurant        Baseball Field   
8             Gastropub                       Café              Bus Line   
9   Japanese Restaurant  Middle Eastern Restaurant        Cosmetics Shop   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0        Discount Store   Distribution Center               Dog Run   
1                 Diner        Discount Store   Distribution Center   
2                  Café               Theater           Yoga Studio   
3            Shoe Store    Miscellaneous Shop              Boutique   
4    Chinese Restaurant    Mexican Restaurant        Sandwich Place   
5                   NaN                   NaN                   NaN   
6        Discount Store   Distribution Center               Dog Run   
7        Discount Store   Distribution Center               Dog Run   
8  Fast Food Restaurant          Intersection                  Bank   
9                  Café                 Diner            Restaurant   

  9th Most Common Venue 10th Most Common Venue  Cluster Labels  
0      Doner Restaurant              Drugstore             1.0  
1               Dog Run       Doner Restaurant             0.0  
2           Event Space                  Hotel             0.0  
3         Women's Store              Gift Shop             0.0  
4      Sushi Restaurant              Nightclub             0.0  
5                   NaN                    NaN             NaN  
6      Doner Restaurant              Drugstore             3.0  
7      Doner Restaurant             Donut Shop             0.0  
8    Athletics & Sports               Pharmacy             0.0  
9               Theater       Ramen Restaurant             0.0

In [36]:
toronto_merged['Cluster Labels'].isnull().sum()
print(toronto_merged.shape)
# dropping rows with  null values 
toronto_merged.dropna(axis = 0, inplace = True) 
print(toronto_merged.shape)


(103, 16)
(99, 16)


Finally, let's visualize the resulting clusters

In [56]:
x = np.arange(kclusters)
print(x)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]
rainbow

[0 1 2 3 4]


['#8000ff', '#00b5eb', '#80ffb4', '#ffb360', '#ff0000']

In [70]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    #print(cluster)
    cluster = int(cluster)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

<a id='item5'></a>

## 5. Examine Clusters

Now, you can examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, you can then assign a name to each cluster. I will leave this exercise to you.

#### Cluster 1

In [71]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0 , toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough 1st Most Common Venue      2nd Most Common Venue  \
1          North York          Hockey Arena               Intersection   
2    Downtown Toronto           Coffee Shop                       Park   
3          North York        Clothing Store     Furniture / Home Store   
4    Downtown Toronto           Coffee Shop                       Park   
7          North York  Gym / Fitness Center        Japanese Restaurant   
8           East York           Pizza Place       Gym / Fitness Center   
9    Downtown Toronto           Coffee Shop             Clothing Store   
10         North York           Pizza Place                       Park   
12        Scarborough                   Bar             History Museum   
13         North York                   Gym           Asian Restaurant   
14          East York                   Spa           Asian Restaurant   
15   Downtown Toronto           Coffee Shop                       Café   
16               York          Hockey Arena                      Field   
17          Etobicoke           Pizza Place          Convenience Store   
18        Scarborough                   Spa          Electronics Store   
19       East Toronto                 Trail          Health Food Store   
20   Downtown Toronto           Coffee Shop               Cocktail Bar   
22        Scarborough           Coffee Shop                   Pharmacy   
23          East York           Coffee Shop        Sporting Goods Shop   
24   Downtown Toronto           Coffee Shop         Italian Restaurant   
25   Downtown Toronto         Grocery Store                       Café   
26        Scarborough                Lounge                     Bakery   
27         North York           Golf Course                       Pool   
28         North York           Coffee Shop                       Bank   
29          East York     Indian Restaurant                Yoga Studio   
30   Downtown Toronto           Coffee Shop                 Restaurant   
31       West Toronto                Bakery                   Pharmacy   
32        Scarborough            Playground           Business Service   
33         North York        Clothing Store                Coffee Shop   
34         North York    Miscellaneous Shop             Massage Studio   
36   Downtown Toronto           Coffee Shop                   Aquarium   
37       West Toronto                   Bar                Coffee Shop   
38        Scarborough            Hobby Shop                Coffee Shop   
39         North York   Japanese Restaurant                       Café   
41       East Toronto      Greek Restaurant                Coffee Shop   
42   Downtown Toronto           Coffee Shop                       Café   
43       West Toronto                  Café             Breakfast Spot   
44        Scarborough                Bakery                   Bus Line   
46         North York         Grocery Store              Shopping Mall   
47       East Toronto                  Park                Coffee Shop   
48   Downtown Toronto           Coffee Shop                 Restaurant   
50         North York           Pizza Place        Empanada Restaurant   
51        Scarborough   American Restaurant                      Motel   
53         North York      Business Service                 Food Truck   
54       East Toronto                  Café                Coffee Shop   
55         North York        Sandwich Place         Italian Restaurant   
56               York    Turkish Restaurant                 Restaurant   
57         North York           Fabric Shop             Baseball Field   
58        Scarborough       College Stadium               Skating Rink   
59         North York      Ramen Restaurant                       Café   
60         North York         Grocery Store             Discount Store   
63               York         Grocery Store                   Bus Line   
65        Scarborough     Indian Restaurant                    Brewery   
67    Central Toro

#### Cluster 2

In [49]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough 1st Most Common Venue 2nd Most Common Venue  \
0         North York                  Park     Food & Drink Shop   
21              York                  Park         Women's Store   
35         East York     Convenience Store           Coffee Shop   
40        North York               Airport                  Park   
49        North York                Bakery                  Park   
61   Central Toronto              Bus Line                  Park   
64              York                  Park     Convenience Store   
66        North York                  Bank                  Park   
83   Central Toronto            Playground                  Park   
85       Scarborough            Playground                  Park   
91  Downtown Toronto                  Park            Playground   
98         Etobicoke                 River                  Park   

         3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                  Yoga Studio            Donut Shop                 Diner   
21                      Market           Yoga Studio            Donut Shop   
35                        Park             Drugstore                 Diner   
40                 Yoga Studio             Drugstore                 Diner   
49  Construction & Landscaping           Yoga Studio             Drugstore   
61                 Swim School           Yoga Studio      Doner Restaurant   
64                 Yoga Studio             Drugstore                 Diner   
66           Convenience Store           Yoga Studio             Drugstore   
83                  Restaurant          Tennis Court           Yoga Studio   
85                  Donut Shop    Dim Sum Restaurant                 Diner   
91                       Trail           Yoga Studio               Dog Run   
98                 Yoga Studio               Dog Run          Dessert Shop   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0         Discount Store   Distribution Center               Dog Run   
21                 Diner        Discount Store   Distribution Center   
35        Discount Store   Distribution Center               Dog Run   
40        Discount Store   Distribution Center               Dog Run   
49        Discount Store   Distribution Center               Dog Run   
61                 Diner        Discount Store   Distribution Center   
64        Discount Store   Distribution Center               Dog Run   
66        Discount Store   Distribution Center               Dog Run   
83   Distribution Center      Department Store          Dessert Shop   
85        Discount Store   Distribution Center               Dog Run   
91          Dessert Shop    Dim Sum Restaurant                 Diner   
98    Dim Sum Restaurant                 Diner        Discount Store   

   9th Most Common Venue 10th Most Common Venue  Cluster Labels  
0       Doner Restaurant              Drugstore             1.0  
21               Dog Run       Doner Restaurant             1.0  
35      Doner Restaurant             Donut Shop             1.0  
40      Doner Restaurant             Donut Shop             1.0  
49      Doner Restaurant             Donut Shop             1.0  
61               Dog Run             Donut Shop             1.0  
64      Doner Restaurant             Donut Shop             1.0  
66      Doner Restaurant             Donut Shop             1.0  
83    Dim Sum Restaurant                  Diner             1.0  
85      Doner Restaurant              Drugstore             1.0  
91        Discount Store    Distribution Center             1.0  
98   Distribution Center       Doner Restaurant             1.0

In [ ]:
cluster_name(toronto_merged.loc[toronto_merged['Cluster Labels'] == 1 ] ) 

#### Cluster 3

In [50]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough 1st Most Common Venue 2nd Most Common Venue  \
45  North York             Cafeteria            Donut Shop   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
45    Dim Sum Restaurant                 Diner        Discount Store   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
45   Distribution Center               Dog Run      Doner Restaurant   

   9th Most Common Venue 10th Most Common Venue  Cluster Labels  
45           Yoga Studio           Dessert Shop             2.0

In [ ]:
cluster_name(toronto_merged.loc[toronto_merged['Cluster Labels'] == 2 ] ) 

#### Cluster 4

In [52]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough 1st Most Common Venue 2nd Most Common Venue  \
6  Scarborough  Fast Food Restaurant           Yoga Studio   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
6            Donut Shop    Dim Sum Restaurant                 Diner   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
6        Discount Store   Distribution Center               Dog Run   

  9th Most Common Venue 10th Most Common Venue  Cluster Labels  
6      Doner Restaurant              Drugstore             3.0

In [ ]:
cluster_name(toronto_merged.loc[toronto_merged['Cluster Labels'] == 3 ] ) 

#### Cluster 5

In [ ]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

In [ ]:
cluster_name(toronto_merged.loc[toronto_merged['Cluster Labels'] == 4 ] ) 

## Name each cluster based on most common Venues in this cluster

In [54]:
#this function will count Most Common Venues 1--> 10 
def cluster_name(data):
    ven = []
    for i in range(4,14): # from 1st Most Common Venue to 10th  Most Common Venue  
        ven= ven+ list(data[data.columns[i]]) # add all Venues in one list  
    ven = pd.DataFrame({'Venue':ven})
    return ven.Venue.value_counts().head() # return most 5 Venues

In [55]:
#dislay 
for i in range(0,5):
    # return most common Venue 
    print('Cluster_No:',i)
    print(cluster_name(toronto_merged.loc[toronto_merged['Cluster Labels'] == i ] ) )
    print()

Cluster_No: 0
Coffee Shop       45
Café              30
Discount Store    26
Restaurant        25
Diner             22
Name: Venue, dtype: int64

Cluster_No: 1
Park                   12
Distribution Center    11
Dog Run                11
Discount Store         11
Yoga Studio            10
Name: Venue, dtype: int64

Cluster_No: 2
Donut Shop            1
Doner Restaurant      1
Dog Run               1
Cafeteria             1
Dim Sum Restaurant    1
Name: Venue, dtype: int64

Cluster_No: 3
Doner Restaurant        1
Discount Store          1
Fast Food Restaurant    1
Dog Run                 1
Dim Sum Restaurant      1
Name: Venue, dtype: int64

Cluster_No: 4
Donut Shop            1
Doner Restaurant      1
Discount Store        1
Dog Run               1
Dim Sum Restaurant    1
Name: Venue, dtype: int64



**Conclusion**:
we can name cluster from most two/three Venues as followig:
- Cluster_0: Coffee shops area
- Cluster_1: Park  and Dance area 
- Cluster_2: Miscellaneous (as all Venues not repated  )  
- Cluster_3: Department Store ,Dance Studio ,Dessert Shop  
- Cluster_4: Department Store ,Dance Studio ,Playground           